sycl/doc/extensions/proposed/sycl_ext_oneapi_group_load_store_extended.asciidoc

```diff
@@ -0,0 +1,220 @@
+
+template <typename T, access::address_space Space, access::decorated IsDecorated>
+void store(multi_ptr<T, Space, IsDecorated> dst, const vec<T, N>& x);
+----
+
+== Extended Proposal
+
+The extended APIs provides the following features:
+
+* load/store with temporary memory buffer, which can be passed via GroupHelper such as `group_with_scratchpad` link:https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_group_sort.asciidoc#group-helper[group_with_scratchpad description from sort over group proposal]
+
+* load/store flexible amount of elements per work item as sycl::span
+
+* specify data placement type via properties (1) or group_memory_helper (*naming is TBD) class (5)
+
+* extra options such as setting boundry values or limiting numbers of work items can be also specified via group_memory_helper
+
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+
+// Load API
+template<typename GroupHelper, typename InputIterator, typename OutputT,
+        std::size_t ElementsPerWorkItem, typename Properties = properties<>>
+void load_over_group(GroupHelper gh, InputIteratorT in_ptr,
+                      sycl::span<OutputT, ItemsPerWorkItem> out, Properties = {}); (1)
+
+template<typename GroupHelper, typename InputIterator, typename OutputT,
+        std::size_t ElementsPerWorkItem, typename GroupMemoryHelper>
+void load_over_group(GroupHelper gh, InputIteratorT in_ptr,
+                      sycl::span<OutputT, ItemsPerWorkItem> out,
+                      GroupMemoryHelper<typename InputIterator::value_type, ItemsPerWorkItem> memory_helper = {}); (2)
+
+// Store API
+template<typename GroupHelper, typename OutputIterator, typename InputT,
+        std::size_t ElementsPerWorkItem, typename Properties = properties<>>
+void store_over_group(GroupHelper gh, OutputIterator out_ptr,
+                      sycl::span<InputT, ItemsPerWorkItem> in, Properties = {}); (3)
+
+template<typename GroupHelper, typename OutputIterator, typename InputT,
+                  std::size_t ElementsPerWorkItem, typename GroupMemoryHelper>
+void store_over_group(GroupHelper gh, OutputIterator out_ptr,
+                      sycl::span<InputT, ItemsPerWorkItem> in,
+                      GroupMemoryHelper<InputT, ItemsPerWorkItem> memory_helper = {}); (4)
+
+// Properties:
+enum class group_algorithm_data_placement {
+  blocked,
+  striped
+};
+
+// Group Memory Helper (aka Loader/Storer) TBD:
+template<typename T, std::size_t ElementsPerWorkItem>
+class group_memory_helper {
+public:
+  template<typename Properties>
+  group_memory_helper(Properties properties = {});
+
+  void set_valid_items_num(std::size_t valid_items);
+
+  void set_out_of_boundry_value(T oob_default);
+
+  static constexpr std::size_t memory_required(sycl::memory_scope scope, std::size_t block_size);
+}; (5)
+
+}
+----
+
+(1) _Preconditions_: `in_ptr` must be the same for all work-items
+in the group.
```

Comment:
Because of this restriction, I think we should rename `load_over_group` to `joint_load` and `store_over_group` to `joint_store`.  This is more aligned with the naming convention of SYCL 2020 group algorithms (e.g. `joint_reduce` requires all work-items to provide the same input range to reduce), and I think it's a more accurate description (because conceptually, the work-items in the group are cooperating to perform a single load/store).

Reply:
Resolved; these have been renamed.

Reply:
@aelizaro - Sorry for reopening this comment, but when I was reading back through again, I realized that there might be some ambiguity here.  I think I got confused and steered you in the wrong direction.

The description of "single value" load says:
```c++
template<typename Group, typename InputIteratorT,
         typename OutputT, typename Properties = properties<>>
void joint_load(Group g, InputIteratorT in_ptr, OutputT& out, Properties = {})
```

> _Effects_: Loads single element from `in_ptr` to `out`.

Are all the work-items loading the same (single) value from `in_ptr` (i.e. broadcasting a result from memory)?  Or are all work-items collaborating to load a contiguous block, and returning one element from the block to each work-item?

If the latter, I think the description of _Effects_ needs to be extended to say that a different value is returned to each work-item.  I also think that I was wrong to suggest `joint_load`/`store`; if a different value is returned to each work-item, the naming should be `group_load`/`store`.

Reply:
the latter, I will rename it back